### DISCUSSION TITLE
Unclosed AsyncSession causes RuntimeError when program exits

### DISCUSSION THREAD

**Discussion by seifertm** [sqlalchemy/sqlalchemy#9235] [Usage Questions]
### Describe the bug

When the user forgets to close an AsyncSession they will encounter a RuntimeError when the AsyncSession is cleaned up.

Although this is clearly a usage error, I would expect that either:
1. The error message in the exception states that there is an unclosed session
2. or the program terminates with a with a warning instead of an exception

The stack trace also warns about `Transaction.rollback` never being run, but I cannot tell if this can be an issue.

### To Reproduce

```python
import asyncio

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession as Session
from sqlalchemy.ext.asyncio import create_async_engine, async_sessionmaker
from sqlalchemy.pool import NullPool


async def main():
    session = SessionLocal()
    result = await session.execute(select(1))
    # User forgets to close the session
    #await session.close()
    return result


if __name__ == "__main__":
    engine = create_async_engine("postgresql+asyncpg://python:changethis@localhost:5432/fastapi", poolclass=NullPool)
    SessionLocal = async_sessionmaker(bind=engine)
    result = asyncio.run(main())
    assert result.first() == (1,)
```

### Error

```
Exception during reset or similar
Traceback (most recent call last):
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 770, in rollback
    self.await_(self._transaction.rollback())
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 93, in await_only
    raise exc.MissingGreenlet(
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/20/xd2s)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 983, in _finalize_fairy
    fairy._reset(
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 1422, in _reset
    pool._dialect.do_rollback(self)
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 657, in do_rollback
    dbapi_connection.rollback()
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 772, in rollback
    self._handle_exception(error)
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 707, in _handle_exception
    exception_mapping = self.dbapi._asyncpg_error_translate
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 1138, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 873, in _asyncpg_error_translate
    import asyncpg
ImportError: sys.meta_path is None, Python is likely shutting down
Exception closing connection <AdaptedConnection <asyncpg.connection.Connection object at 0x7f38e375b120>>
Traceback (most recent call last):
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 770, in rollback
    self.await_(self._transaction.rollback())
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 93, in await_only
    raise exc.MissingGreenlet(
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/20/xd2s)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 983, in _finalize_fairy
    fairy._reset(
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 1422, in _reset
    pool._dialect.do_rollback(self)
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 657, in do_rollback
    dbapi_connection.rollback()
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 772, in rollback
    self._handle_exception(error)
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 707, in _handle_exception
    exception_mapping = self.dbapi._asyncpg_error_translate
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 1138, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 873, in _asyncpg_error_translate
    import asyncpg
ImportError: sys.meta_path is None, Python is likely shutting down

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 380, in _close_connection
    self._dialect.do_terminate(connection)
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 984, in do_terminate
    dbapi_connection.terminate()
  File "/tmp/venv/lib/python3.9/site-packages/sqlalchemy/dialects/postgresql/asyncpg.py", line 793, in terminate
    self._connection.terminate()
  File "/tmp/venv/lib/python3.9/site-packages/asyncpg/connection.py", line 1344, in terminate
    self._abort()
  File "/tmp/venv/lib/python3.9/site-packages/asyncpg/connection.py", line 1371, in _abort
    self._protocol.abort()
  File "asyncpg/protocol/protocol.pyx", line 571, in asyncpg.protocol.protocol.BaseProtocol.abort
  File "/usr/lib/python3.9/asyncio/selector_events.py", line 678, in abort
    self._force_close(None)
  File "/usr/lib/python3.9/asyncio/selector_events.py", line 729, in _force_close
    self._loop.call_soon(self._call_connection_lost, exc)
  File "/usr/lib/python3.9/asyncio/base_events.py", line 751, in call_soon
    self._check_closed()
  File "/usr/lib/python3.9/asyncio/base_events.py", line 515, in _check_closed
    raise RuntimeError('Event loop is closed')
RuntimeError: Event loop is closed
/tmp/venv/lib/python3.9/site-packages/sqlalchemy/pool/base.py:1022: RuntimeWarning: coroutine 'Transaction.rollback' was never awaited
```
### Versions

- OS: Linux
- Python: 3.9, 3.10
- SQLAlchemy: 2.0.1
- Database: PostgreSQL 15.1
- DBAPI (eg: psycopg, cx_oracle, mysqlclient): asyncpg==0.27.0

### Additional context

Here's the Docker command that was used to spin up the database:
```
$ docker run --rm -p 5432:5432 --name some-postgres -e POSTGRES_USER=python -e POSTGRES_PASSWORD=changethis -e POSTGRES_DB=fastapi -d postgres
```

**Comment by CaselIT** [comment]
Hi,

The runtime expection here is caused by the even loop closed and the interpreter shutting down.
I'm not really shure what you expect sqlalchemy to do here. These are all conditions that are outside the library control

**Reply by seifertm** [reply]
Sqlalchemy has no control over the interpreter shutting down. We fully agree on that. 

I filed the issue for two reasons:
1. The resulting stack trace doesn't provide any hints that would help the user find out what's wrong with their code
2. It looks like sqlalchemy is trying to clean up, *despite* the fact that the interpreter is shutting down

There's precendent in the standard library for addressing both points. For example, the "missing user information" is solved by emitting a ResourceWarning when leaving file handles or event loops unclosed.

The Unix event loop implementation uses `sys.is_finalizing` to check whether the interpreter is shutting down or not.

If you say this cannot be solved by sqlalchemy, I can totally accept that. I just believe it would be worthwhile to try and exit gracefully or at least introduce a dedicated warning or error message specifically for this case. Just to let the user know they forgot to close a session.

**Reply by zzzeek** [reply]
> Sqlalchemy has no control over the interpreter shutting down. We fully agree on that.
> 
> I filed the issue for two reasons:
> 
>     1. The resulting stack trace doesn't provide any hints that would help the user find out what's wrong with their code

I'm not aware of any techniques to affect the shape of a stack trace in the case that the interpreter is shutting down.


> 
>     2. It looks like sqlalchemy is trying to clean up, _despite_ the fact that the interpreter is shutting down

interpreter shutdown is completely chaotic so I really dont think it would be a good idea for us to have a conditional which states "guess if the interpreter is actually shutting down and then don't do any cleanup".    this would be a major increase of risk.

> 
> 
> There's precendent in the standard library for addressing both points. For example, the "missing user information" is solved by emitting a ResourceWarning when leaving file handles or event loops unclosed.

the "cleanup" that's happening here is that the connection pool attempts to emit "rollback" on a connection when its returned to the pool.   In this test case, it is being triggered by gc weakref callback, which is a positive indicator that the connection was not returned to the pool.

Not too sure how to use ResourceWarning, per https://docs.python.org/3/library/exceptions.html?highlight=resourcewarning#ResourceWarning it says it's ignored by default.

If I add this warning, I see nothing in the output:

```diff
diff --git a/lib/sqlalchemy/pool/base.py b/lib/sqlalchemy/pool/base.py
index a77f652050..7b6018fdb2 100644
--- a/lib/sqlalchemy/pool/base.py
+++ b/lib/sqlalchemy/pool/base.py
@@ -30,6 +30,8 @@ from typing import TYPE_CHECKING
 from typing import Union
 import weakref
 
+import warnings
+
 from .. import event
 from .. import exc
 from .. import log
@@ -945,6 +947,11 @@ def _finalize_fairy(
         _strong_ref_connection_records.pop(weakref.ref(fairy), None)
 
     if is_gc_cleanup:
+
+        warnings.warn(
+            "Pool connection was not returned to the pool explicitly",
+            ResourceWarning,
+        )
         assert connection_record is not None
         if connection_record.fairy_ref is not ref:
             return
```

so not sure how to make ResourceWarning useful.   If I put a normal warning there, then we do see it, but it's quite buried amongst all the stack traces.     We had considered turning off this gc-based cleanup for 2.0 entirely however a lot of reasons came up (which I dont recall offhand) to keep it in, not sure if adding a normal warning here affects those use cases.   2.0 is now released so adding a new foundational warning is probably too late (could be in 2.1, but getting best practices and precedent would be important here - I dont think for example asyncpg itself emits any warning?). 


> 
> The Unix event loop implementation uses `sys.is_finalizing` to check whether the interpreter is shutting down or not.

That is actually news to me but again I really dont think it's a good idea to add logic here to *not* do normal cleanup.   It looks like you had an incorrectly written asyncio program and got confused, I am very sorry about that, but asyncio has a lot of difficult edges to it.     A warning for pool connection not explicitly returned is doable but I'd want to follow best practices.

**Reply by CaselIT** [reply]
Also note that sqlalchemy initiates the rollback, but the the asyncio error is triggered by asyncpg, that is trying to schedule something on a closed loop.

There may be a problem with sqlalchemy in any case, since it's trying to trigger a rollback without the asyncio greenlet. I thought we changed that to not do that. Or am I misremembering mike?

**Reply by seifertm** [reply]
Thanks for your explanations! I can see that not trying to rollback a connection is a potential risk.

> I'm not aware of any techniques to affect the shape of a stack trace in the case that the interpreter is shutting down.

The idea was to check whether an event loop was running before submitting coroutines and to raise custom exception when no loop is running.


>  It looks like you had an incorrectly written asyncio program and got confused, I am very sorry about that

I wasn't directly affected by this issue. I investigated a [pytest-asyncio bug report](https://github.com/pytest-dev/pytest-asyncio/issues/491) and I thought sharing the results could be useful. I'm sorry if this created more noise than it's worth.

As for the ResourceWarning: They're disabled by default, but can be turned on with the appropriate [command-line option](https://docs.python.org/3/using/cmdline.html#cmdoption-W) (e.g. `-W default`). There's a possibility to turn warnings into errors with `-W error` or perform [more complex filtering](https://docs.python.org/3/library/warnings.html#describing-warning-filters). I believe that the interpreter can even trace back the line where an object was allocated, if that object is passed in the `source` kwarg of ResourceWarning and [tracemalloc](https://docs.python.org/3/library/tracemalloc.html) is enabled. The docs on [Python development mode](https://docs.python.org/3/library/devmode.html#python-development-mode) provide some examples for ResourceWarnings.

**Reply by CaselIT** [reply]
> > I'm not aware of any techniques to affect the shape of a stack trace in the case that the interpreter is shutting down.
> 
> The idea was to check whether an event loop was running before submitting coroutines and to raise custom exception when no loop is running.

The thing is that here sqlalchemy is _not_ calling a coroutine. It's calling [`terminate`](https://github.com/MagicStack/asyncpg/blob/7df9812a068c95e5dd4aa1d0270db8f177ee1e50/asyncpg/connection.py#L1341-L1345), so that particular check should be in asyncpg if it uses the loop on a sync call.

That said the MissingGreenly is sqlalchemy doing.

I'll try taking a look why it's calling rollback in a non greenlet context here

**Reply by zzzeek** [reply]
we should not be calling rollback if we are in a gc handler, there is an explicit flag set up in _finalize_fairy that checks for gc'ed, however there may be a slight flaw in this logic since we added support for asyncpg terminate, so yes that might be an actual bug here


```py
    dont_restore_gced = (
        pool._dialect.is_async and not pool._dialect.has_terminate
    )

    if dont_restore_gced:
        detach = connection_record is None or is_gc_cleanup
        can_manipulate_connection = ref is None
    else:
        detach = connection_record is None
        can_manipulate_connection = True

```

for asyncpg we set dont_restore_gc'ed to false since we can call the terminate command.  BUT then we do the reset with can_manipulate_connection = True and that is incorrect.    The logic should be broken out into two separate flags, "can_terminate_connection" and "can_manipulate_connection".   check out the code and we can open an issue for that.  will likely make this particular stack smoother.

in fact yeah if we make this change just to test:

```diff
diff --git a/lib/sqlalchemy/pool/base.py b/lib/sqlalchemy/pool/base.py
index a77f652050..7e514acb41 100644
--- a/lib/sqlalchemy/pool/base.py
+++ b/lib/sqlalchemy/pool/base.py
@@ -953,7 +953,7 @@ def _finalize_fairy(
 
     # null pool is not _is_asyncio but can be used also with async dialects
     dont_restore_gced = (
-        pool._dialect.is_async and not pool._dialect.has_terminate
+        pool._dialect.is_async #and not pool._dialect.has_terminate
     )
 
     if dont_restore_gced:

```

test program above output:

```
The garbage collector is trying to clean up connection <AdaptedConnection <asyncpg.connection.Connection object at 0x7f0a76f5f840>>. This feature is unsupported on asyncio dbapis that lack a "terminate" feature, since no IO can be performed at this stage to reset the connection. Please close out all connections when they are no longer used, calling ``close()`` or using a context manager to manage their lifetime.
sys:1: SAWarning: The garbage collector is trying to clean up connection <AdaptedConnection <asyncpg.connection.Connection object at 0x7f0a76f5f840>>. This feature is unsupported on asyncio dbapis that lack a "terminate" feature, since no IO can be performed at this stage to reset the connection. Please close out all connections when they are no longer used, calling ``close()`` or using a context manager to manage their lifetime.

```

how about that?

**Reply by zzzeek** [reply]
see #9237

**Reply by CaselIT** [reply]
that's still not 100% correct, or at least the message is not correct, since asyncpg *does* have terminate, while the are complaining that it doesn't

**Reply by zzzeek** [reply]
Well when we split up the boolean we have to split up the message also, likely the reset method needs its own warning

**Reply by zzzeek** [reply]
I think actually we are supposed to terminate the connection unconditionally if we are async and we are in a gc.  I think the fix to add terminate support was done incorrectly
